CompuCell3D/core/CompuCell3D/Potts3D/Potts3D.cpp

```diff
@@ -735,377 +736,398 @@ unsigned int Potts3D::metropolisFast(const unsigned int steps, const double temp
+                }
+                ++attemptedECVec[currentWorkNodeNumber];
 
-				if (changePixelCell == cell) {
-					continue;//skip the rest of the loop if change pixel points to the same cell as pt
-				}
-				else {
-					;
+                flipNeighborVec[currentWorkNodeNumber] = pt;
 
-				}
+                /// change takes place at change pixel  and pt is a neighbor of changePixel
+                // Calculate change in energy				
 
+                double change = energyCalculator->changeEnergy(changePixel, cell, changePixelCell, i);
 
-				if (sizeFrozenTypeVec && changePixelCell) {///must also make sure that cell ptr is different 0; Will never freeze medium
-					if (checkIfFrozen(changePixelCell->type))
-						continue;
-				}
-				++attemptedECVec[currentWorkNodeNumber];
+                // Acceptance based on probability
+                double motility = fluctAmplFcn->fluctuationAmplitude(cell, changePixelCell);
 
-				flipNeighborVec[currentWorkNodeNumber] = pt;
+                double prob = acceptanceFunction->accept(motility, change);
 
-				/// change takes place at change pixel  and pt is a neighbor of changePixel
-				// Calculate change in energy				
 
-				double change = energyCalculator->changeEnergy(changePixel, cell, changePixelCell, i);
+                if (numberOfThreads == 1) {
+                    energyCalculator->set_aceptance_probability(prob);
+                }
 
-				// Acceptance based on probability
-				double motility = fluctAmplFcn->fluctuationAmplitude(cell, changePixelCell);
+                if (prob >= 1.0 || rand->getRatio() < prob) {
+                    // Accept the change
+                    if (test_output_generate_flag) {
 
-				double prob = acceptanceFunction->accept(motility, change);
-				//#pragma omp critical
-				//				{
-				//				cerr<<" thread="<<currentWorkNodeNumber<<" section="<<s<<" prob="<<prob<<" energy="<<energy<<endl;
-				//				}
+                        PottsTestData potts_test_data;
 
+                        potts_test_data.changePixel = n.pt;
+                        potts_test_data.changePixelNeighbor = pt;
+                        potts_test_data.motility = motility;
+                        potts_test_data.pixelCopyAccepted = true;
+                        potts_test_data.acceptanceFunctionProbability = prob;
+                        if (connectivityConstraint) {
+                            potts_test_data.using_connectivity = true;
+                            potts_test_data.connectivity_energy = connectivityConstraint->changeEnergy(changePixel, cell, changePixelCell);
+                            if (potts_test_data.connectivity_energy) {
+                                potts_test_data.pixelCopyAccepted = false;
+                            }
+                        }
 
-				if (numberOfThreads == 1) {
-					energyCalculator->set_aceptance_probability(prob);
-				}
+                        energyCalculator->log_output(potts_test_data);

```

Comment:
A useful but unnecessary feature that's probably beyond the scope of this PR might be to add support for testing using all Metropolis implementations.

Reply:
Good point, I was thinking about it too but current tests focus on energy functions and (indirectly) lattice monitors. As you can see the test data generation mode records pixel copy neighbors, energy function calculations and then those are "replayed" in a separate function that is independent from various metropolis functions. So one thing we could do (separate PR ) is to add test data generation to metropolisBoundaryWalker and maybe seriously consider removing metrolpolisList function (if it still exists...)